### PR TITLE
Input support custom event target.

### PR DESCRIPTION
**Pull request by cptbtptpbcptdtptp**

### CODE REVIEW

File: packages/core/src/input/wheel/WheelManager.ts [hunk 1]
```diff
@@ -11,19 +11,28 @@ export class WheelManager implements IInput {
...
+
+  set target(value: EventTarget) {
+    const preTarget = this._target;
+    if (preTarget !== value) {
+      preTarget && this._removeEventListener();
+      this._target = value;
+      value && this._addEventListener();
+    }
+  }
+
```
Comment (GuoLei1990) [inline packages/core/src/input/wheel/WheelManager.ts:29]: Don't need open this， Open this need do some warp in cross platform engine
The historical baggage is large and there is no urgent need for the time being.

File: packages/core/src/input/wheel/WheelManager.ts [hunk 2]
```diff
@@ -11,19 +11,28 @@ export class WheelManager implements IInput {
...
   /**
    * Create a KeyboardManager.
    */
-  constructor(engine: Engine) {
-    // @ts-ignore
-    const htmlCanvas = engine._canvas._webCanvas;
+  constructor(engine: Engine, target: EventTarget = window) {
```
Comment (GuoLei1990) [inline packages/core/src/input/wheel/WheelManager.ts:33]: Add @internal in comments

File: packages/core/src/input/wheel/WheelManager.ts [hunk 3]
```diff
@@ -70,12 +79,8 @@ export class WheelManager implements IInput {
    * @internal
    */
   _destroy(): void {
-    if (this._hadListener) {
-      this._canvas.removeEventListener("wheel", this._onWheelEvent);
-      this._hadListener = false;
-    }
+    this.target = null;
```
Comment (GuoLei1990) [inline packages/core/src/input/wheel/WheelManager.ts:82]: Maybe `_removeEventListener`

File: packages/core/src/Engine.ts [hunk 4]
```diff
@@ -754,4 +754,10 @@ export interface EngineConfiguration {
   colorSpace?: ColorSpace;
   /** Shader lab */
   shaderLab?: IShaderLab;
+  /** Input options */
```
Comment (GuoLei1990) [inline packages/core/src/Engine.ts:757]: Irregular comments

File: packages/core/src/Engine.ts [hunk 5]
```diff
@@ -754,4 +754,10 @@ export interface EngineConfiguration {
...
+  inputOptions?: {
```
Comment (GuoLei1990) [inline packages/core/src/Engine.ts:758]: `config.inputOptions.xx` or `config.input.xx`?

File: packages/core/src/Engine.ts [hunk 6]
```diff
@@ -754,4 +754,10 @@ export interface EngineConfiguration {
...
+    keyboard?: EventTarget;
+    pointer?: EventTarget;
+    wheel?: EventTarget;
```
Comment (GuoLei1990) [inline packages/core/src/Engine.ts:761]: `keyboard`、`pointer`、`wheel` is not clear  

`keyboardTarget` .... maybe is better

File: packages/core/src/input/InputManager.ts [hunk 7]
```diff
@@ -150,18 +150,14 @@ export class InputManager {
   /**
    * @internal
    */
-  constructor(engine: Engine) {
+  constructor(engine: Engine, inputOptions?: { keyboard?: EventTarget; pointer?: EventTarget; wheel?: EventTarget }) {
```
Comment (GuoLei1990) [inline packages/core/src/input/InputManager.ts:153]: inputConfiguration

File: packages/core/src/input/pointer/PointerManager.ts [hunk 8]
```diff
@@ -36,26 +36,34 @@ export class PointerManager implements IInput {
...
 
-  /**
-   * Create a PointerManager.
-   * @param engine - The current engine instance
-   * @param htmlCanvas - HTMLCanvasElement
-   */
-  constructor(engine: Engine) {
-    // @ts-ignore
-    const htmlCanvas = engine._canvas._webCanvas;
+  get target(): EventTarget {
```
Comment (GuoLei1990) [inline packages/core/src/input/pointer/PointerManager.ts:45]: The same with wheel manager

File: packages/core/src/Engine.ts (hunk 6 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/Engine.ts]: Done.

File: packages/core/src/input/pointer/PointerManager.ts (hunk 8 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/input/pointer/PointerManager.ts]: Done.

File: packages/core/src/input/keyboard/KeyboardManager.ts [hunk 9]
```diff
@@ -25,24 +25,32 @@ export class KeyboardManager implements IInput {
...
-   * Create a KeyboardManager.
-   */
-  constructor(engine: Engine) {
-    // @ts-ignore
-    const htmlCanvas = engine._canvas._webCanvas;
+  get target(): EventTarget {
+    return this._target;
+  }
+
+  set target(value: EventTarget) {
```
[the remark this answers is not in the document]
Comment (cptbtptpbcptdtptp) [reply packages/core/src/input/keyboard/KeyboardManager.ts]: Done.

File: packages/core/src/input/wheel/WheelManager.ts (hunk 3 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/input/wheel/WheelManager.ts]: Done.

File: packages/core/src/input/wheel/WheelManager.ts (hunk 2 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/input/wheel/WheelManager.ts:20]: Done.

File: packages/core/src/input/wheel/WheelManager.ts (hunk 1 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/input/wheel/WheelManager.ts:16]: Done.

**Review by GuoLei1990 — APPROVED** [review]

**Pull request merged**